packages/react-components/src/components/MessageThread.tsx

```diff
@@ -466,6 +474,10 @@ export type MessageThreadProps = {
    * @param messageStatusIndicatorProps - props of type MessageStatusIndicatorProps
    */
   onRenderMessageStatus?: (messageStatusIndicatorProps: MessageStatusIndicatorProps) => JSX.Element | null;
+  /**
+   * A callback function that can be used to provide custom data to an Avatar.
+   */
+  avatarPersonaDataProvider?: AvatarPersonaDataProvider;
```

Comment:
I 100% agree that this should be exposed in composite level(and that's a smart design! Love the plain obj we return for userId)

But I am not sure whether we should expose this in component level - in component level, we are asking skilled engineers to do their own stuff, and this should scenario could be covered by onRenderAvatar

I guess we could do something like this

1. pass PersonaDataProvider to composite
2. In composite, implement an onRenderAvatar function and call personaDataProvider internally
3. pass onRenderAvatar function down to the component

It might fit our concept of the boundary between component and composite better - 
composite is something we prebuild for developers, just provide us {[key]: value}, composite will render things for you (maybe we can even remove onRenderAvatar function from composite if this design is powerful enough),

but component - if you choose this path, you are a skilled engineer, and wanna to have more control, only override color of persona is not enough(you could do it by onRenderAvatar still though), then we provide you onRenderAvatar api to do anything you want

Reply:
Thanks for reviewing the PR and your comments @PorterNan 
If we use an `onRender` function, the issue will arise when Components want to change the size of Personas.
If composites provide the Persona components to the underlying components using `onRender`, any design change in Persona becomes really painful and difficult to test.

By allowing only primitive data modifications, we ensure that our design system is not broken by CONTOSO or `onRender` functions.

Reply:
I like to think of the components independent of the composites, first.

So, 
- `AvatarPersona` should simply take static props.
- Complex components, like `ParticipantPane` that need to show the avatar for a dynamic number of users, should provide a way to inject `AvatarPersona`, i.e. `(userId: string) => Promise<AvatarPersona>`.
- Composites, should collect together just the data that end-users might want to give us, i.e., `(userId: string) => Promise<AvatarPersonaData>`.

Also, related to one of my comments above, `AvatarPersonaProps` should follow the practices of other props -- arbitrary styling via `styles` prop etc. `AvatarPersonaData` on the other hand bundles together just the things a user of the composite wants (e.g. `initialsTextColor`).

Reply:
@prprabhu-ms 
1. Any particular reason why `AvatarPersona` should only take static props?
2. This means duplication of code everywhere `AvatarPersona` is used. Specifically, the code for  handling and fetching the data through CONTOSO provided provider function. That also defeats the purpose of wrapping `Persona` component inside a custom `AvatarPersona` components whose sole purpose is to reduce duplication of common code that handles fetching of contoso data.
3. That's partially how it is happening in this design. If composite were to fetch the data, it would prematurely fetch the data even before an avatar is rendered. Plus you would have to track the state of userId in composite before you could call this function. More complexity at the composite layer to maintain. When it could have simply be offloaded to a child component `AvatarPersona` which has a single responsiblity.

Using `AvatarPersona` in the current way also means that we are implementing SOLID principles to a large extent.